### ISSUE TITLE
[Bug] Streamline enableAjaxTable() with functionality.

### ISSUE THREAD

**Issue by OwenMelbz** [Laravel-Backpack/CRUD#411]
Lots of people have noticed that when you enableAjax certain functionality breaks.

This includes

- Ordering inc dates
- Model Functions
- Accessors/Mutators/Attributes
- Fake/Aggregated Columns.
- Searching relationships

From now on, if you are having issues similar to this with enableAjax, can we please keep the discussion consolidated here so its easier to keep track of.

For the time being I'll be closing lots of tickets to tidy up, however they'll still be linked to below

https://github.com/Laravel-Backpack/CRUD/issues/407
https://github.com/Laravel-Backpack/CRUD/issues/402
https://github.com/Laravel-Backpack/CRUD/issues/370
https://github.com/Laravel-Backpack/CRUD/issues/357
https://github.com/Laravel-Backpack/CRUD/issues/306
https://github.com/Laravel-Backpack/CRUD/issues/324
https://github.com/Laravel-Backpack/CRUD/issues/183

**Comment by jsiebach** [comment]
In order to effectively do AJAX, the sort/search column has to be generated in MYSQL, right?  Otherwise you would have to load all of the models and then filter or sort after the fact.  This really defeats the purpose of the AJAX table in a lot of cases.

I would think setting specific columns that are able to be used as sort or order columns would be necessary (or deriving that information from the column type).  Attributes, accessors, fake columns... none of these could be used in any situation that I can think of.  Relationships could be searched/ordered through joins in the query... maybe derived columns could be ordered IF there was mysql equivalent.  For instance:

```
public function getFullNameAttribute()
{
  return $this->first_name . " " . $this->last_name
}
```

This could never be sorted on, because it is generated after the query runs, however it is possible to sort by those two columns in a query.  If we could provide an extra parameter to the column, maybe we could pass in a function that generates the `ORDER BY` clause, ie:

```
$this->addColumn[
  ...
  'orderBy' => function ($direction) {
    return "ORDER BY first_name $direction,last_name $direction";
  }
```

Then this could be added to the query in the case that someone tried to sort by this column.

Sorry I don't have more to add at this time, I just wanted to join this discussion and get ping'd when things change.  I have a few places I'd like to use AJAX in my current project but am unable to, and want to help!

**Comment by jsiebach** [comment]
You could also add `'orderBy' => false` and then the column should be non-clickable on the front end

**Comment by adbmdp** [comment]
I'm also facing the same problems but it's too hard for me as i'm just learning Laravel...
I really need the ajax functionalities for my project as i have big tables. But there is too many bugs with enableAjax.

**Comment by MDEvo** [comment]
Just want to share some code that i just made after work around for few days to fix some of the datatables issue :

-  Column with Accessors

- Searching relationships

Others issue you may test it on your project as i only test for my project only. Good Luck!


This code using the Laravel Datatables (https://github.com/yajra/laravel-datatables) 

In order to use this you need to install Laravel Datatables Package

`composer require yajra/laravel-datatables-oracle`

then modify AjaxTable.php or add in your controller

`  public function search()
  {
      $this->crud->hasAccessOrFail('list');

      // create an array with the names of the searchable columns
      $columns = collect($this->crud->columns)
                  ->reject(function ($column, $key) {
                      // the select_multiple columns are not searchable
                      return isset($column['type']) && $column['type'] == 'select_multiple';
                  })
                  ->pluck('name')
                  // add the primary key, otherwise the buttons won't work
                  ->merge($this->crud->model->getKeyName())
                  ->toArray();

      //identify non text columns
      $columns_type_not_text = collect($this->crud->columns)->filter(function ($column) {
          return isset($column['type']) && $column['type'] != 'text';
        })->pluck('name')->toArray();

      $data = $this->crud->getEntries(); // $this->crud->query->get();

      $datatable_data = collect();
      foreach ($data as $dtrow) {
          $new_dtrow = clone $dtrow; 

          foreach($dtrow->toArray() as $k => $v){
            if( !in_array($k, $columns) ){ // remove unrelated column data
              unset($new_dtrow->$k);
            }
          }

          // append additional data e.g. accesor / relation data for search function
          foreach ($columns as $key => $item) {
            if( in_array($item, $columns_type_not_text) ){  //relation
                $col_prop = collect($this->crud->columns)->where('name', $item)->first();
                $entity = isset($col_prop['entity']) ? $col_prop['entity'] : null;
                $attribute = isset($col_prop['attribute']) ? $col_prop['attribute'] : null;

                if($entity != null && $attribute != null){
                  $new_dtrow->{'srch_'.$item} = $dtrow->{$entity}[$attribute];
                }
            }

            if( !array_key_exists($item, $dtrow) ){  //accesor
                $new_dtrow->$item = $dtrow->{$item};
            }
          }

          $datatable_data->push($new_dtrow);
      }

      $data_columns = collect($datatable_data->first()->toArray())->keys()->all();

      $datatables = \Yajra\Datatables\Datatables::of($datatable_data);
      $datatables = $datatables->setRowId('id'); //set row id

      foreach($data_columns as $column) {
        $datatables = $datatables->removeColumn($column); //remove unwanted data
      }

      $crud = $this->crud;
      // add the details_row buttons as the first column
      if ($this->crud->details_row) {
          $datatables = $datatables->addColumn(0, function($entry) use ($crud) {
              return \View::make('crud::columns.details_row_button')
                            ->with('crud', $crud)
                            ->with('entry', $entry)
                            ->render();
          });
      }

      // get the actual HTML for each row's cell
      foreach ($this->crud->columns as $key => $column) {
          if( isset($column['type']) && $column['type'] != 'text' ){
              $datatables = $datatables->addColumn(0, function($entry) use ($column) {
                return $this->crud->getCellView($column, $entry);
              });
          } else {
              $datatables = $datatables->addColumn(0, function($entry) use ($column) {
                return $entry->{$column["name"]};
              });
          }
      }

      // add the buttons as the last column
      if ($this->crud->buttons->where('stack', 'line')->count()) {
          $datatables = $datatables->addColumn(0, function($entry) use ($crud) {
              return \View::make('crud::inc.button_stack', ['stack' => 'line'])
                          ->with('crud', $crud)
                          ->with('entry', $entry)
                          ->render();
          });
      }

      $datatables = $datatables->escapeColumns($columns);

      // Override global search function
      $datatables = $datatables->filter(function ($instance) {
          if ($keyword = $this->request->get('search')['value']) {
              $instance->collection = $instance->collection->filter(function ($row) use ($keyword) {
                  $row = $row->toArray();
                  $col_name = array_keys($row);
                  $results = false;
                  foreach($col_name as $col){
                    if ( !is_array($row[$col]) ){
                      $result = str_contains(strtolower($row[$col]), strtolower($keyword));
                      $results = ($result == true) ? $result : $results;
                    }
                  }
                  return $results;

              });
          }

      });

      return $datatables->make(true);

  }`

**Comment by viktorivanov** [comment]
[This approach](https://github.com/Laravel-Backpack/CRUD/issues/402#issuecomment-287318385) might be acceptable for someone wanting to use accessor as column with enableAjaxTable turned on.

**Comment by errogaht** [comment]
dont know how name it.
in model i have
![](https://goo.gl/4GfmSh)
When im use `enableAjaxTable()`
it produce this req ```select `title` as `title`, `category_id` as `category_id`, `id` as `id` from `blog_posts` order by `created_at` desc limit 25 offset 0```
and result is `$this->slug is null` and my function return wrong result.

problem is here `$dataTable = new \LiveControl\EloquentDataTable\DataTable($this->crud->query, $columns);` 
dont know how but query must be 
```select * from `blog_posts` order by `created_at` desc limit 25 offset 0```

**Comment by lloy0076** [comment]
This is closed by #710.
